flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1389,6 +1388,149 @@ private void copyStateDataHandleData(
+			throw new UnsupportedOperationException("The new namespace serializer requires state migration in order for the job to proceed." +
+				" However, migration for state namespace currently isn't supported.");
+		}
+
+		if (stateCompatibility.isRequiresMigration()) {
+			migrateStateValue(stateDesc, stateInfo, restoredMetaInfoSnapshot, newMetaInfo);
+		} else {
+			newMetaInfo = new RegisteredKeyValueStateBackendMetaInfo<>(
+				newMetaInfo.getStateType(),
+				newMetaInfo.getName(),
+				newMetaInfo.getNamespaceSerializer(),
+				stateSerializer,
+				snapshotTransformer);
+		}
+
+		stateInfo.f1 = newMetaInfo;
+		return newMetaInfo;
+	}
+
+	/**
+	 * Migrate only the state value, that is the "value" that is stored in RocksDB. We don't migrate
+	 * the key here, which is made up of key group, key, namespace and map key
+	 * (in case of MapState).
+	 */
+	private <N, S extends State, SV> void migrateStateValue(
+		StateDescriptor<S, SV> stateDesc,
+		Tuple2<ColumnFamilyHandle, RegisteredStateMetaInfoBase> stateInfo,
+		StateMetaInfoSnapshot restoredMetaInfoSnapshot,
+		RegisteredKeyValueStateBackendMetaInfo<N, SV> newMetaInfo) throws Exception {
+
+		if (stateDesc.getType().equals(StateDescriptor.Type.MAP)) {
+			throw new StateMigrationException("The new serializer for a MapState requires state migration in order for the job to proceed." +
+				" However, migration for MapState currently isn't supported.");
+		}
+
+		LOG.info(
+			"Performing state migration for state {} because the state serializer changed in an incompatible way.",
+			stateDesc);
+
+		// we need to get an actual state instance because migration is different
+		// for different state types. For example, ListState needs to deal with
+		// individual elements
+		StateFactory stateFactory = STATE_FACTORIES.get(stateDesc.getClass());
+		if (stateFactory == null) {
+			String message = String.format("State %s is not supported by %s",
+				stateDesc.getClass(), this.getClass());
+			throw new FlinkRuntimeException(message);
+		}
+		State state = stateFactory.createState(
+			stateDesc,
+			Tuple2.of(stateInfo.f0, newMetaInfo),
+			RocksDBKeyedStateBackend.this);
+		if (!(state instanceof AbstractRocksDBState)) {
+			throw new FlinkRuntimeException(
+				"State should be an AbstractRocksDBState but is " + state);
+		}
+
+		@SuppressWarnings("unchecked")
+		AbstractRocksDBState<?, ?, SV, S> rocksDBState = (AbstractRocksDBState<?, ?, SV, S>) state;
+
+		Snapshot rocksDBSnapshot = db.getSnapshot();
+		try (RocksIteratorWrapper iterator = getRocksIterator(db, stateInfo.f0)) {
+
+			iterator.seekToFirst();
+			while (iterator.isValid()) {
+				byte[] serializedValue = iterator.value();
+				byte[] migratedSerializedValue = rocksDBState.migrateSerializedValue(
+					serializedValue,
+					(TypeSerializer<SV>) restoredMetaInfoSnapshot.getTypeSerializerConfigSnapshot(StateMetaInfoSnapshot.CommonSerializerKeys.VALUE_SERIALIZER).restoreSerializer(),
```

Comment:
Why are we doing this inside the loop? I would do this only once before the loop and hold the serializer in a local variable. Furthermore, please suppress the unchecked warning.